crates/llms/src/nql/mod.rs

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::missing_errors_doc)]
+use async_trait::async_trait;
+use snafu::Snafu;
+
+#[cfg(feature = "candle")]
+pub mod candle;
+
+#[cfg(feature = "mistralrs")]
+pub mod mistral;
+
+pub enum LlmRuntime {
+    Mistral,
+    Candle,
+}
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to run the NSQL model"))]
+    FailedToRunModel {
+        e: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Local model, expected at {expected_path}, not found"))]
+    LocalModelNotFound { expected_path: String },
+
+    #[snafu(display("Local tokenizer, expected at {expected_path}, not found"))]
+    LocalTokenizerNotFound { expected_path: String },
+
+    #[snafu(display("Failed to load model from file {e}"))]
+    FailedToLoadModel {
+        e: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Failed to load model tokenizer"))]
+    FailedToLoadTokenizer {
+        e: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Failed to tokenize"))]
+    FailedToTokenize {
+        e: Box<dyn std::error::Error + Send + Sync>,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[async_trait]
+pub trait Nql: Sync + Send {
+    async fn run(&mut self, prompt: String) -> Result<Option<String>>;
+}
+
+/// Loads `https://huggingface.co/motherduckdb/DuckDB-NSQL-7B-v0.1-GGUF` local spice llms cache (i.e. ~/.spice/llms/),
+/// based on the configured features, and the preference order: [`mistralrs`, `candle`].
+#[allow(unreachable_patterns)]
```

Comment:
Is this because of the match arms potentially being exhaustive and thus the catch-all arm `_` isn't needed if all features are enabled?

Reply:
Yes!